src/libstrongswan/plugins/openssl/openssl_kem.c

```diff
@@ -0,0 +1,360 @@
+	chunk_t seed = chunk_empty;
+	drbg_t *drbg = this->thread_drbg->get(this->thread_drbg);
+
+	pkey = EVP_PKEY_kem_new_raw_public_key(openssl_kem_get_nid(this),
+																				 public_key.ptr, public_key.len);
+	if (!pkey)
+	{
+		return FALSE;
+	}
+	ctx = EVP_PKEY_CTX_new(pkey, NULL);
+	if (!ctx)
+	{
+		EVP_PKEY_free(pkey);
+		return FALSE;
+	}
+	if (drbg)
+	{
+		EVP_PKEY_encapsulate_deterministic(ctx, NULL, &ciphertext_length, NULL,
+																			 &shared_secret_length, NULL,
+																			 &seed_length);
+		this->shared_secret = chunk_alloc(shared_secret_length);
+		this->ciphertext = chunk_alloc(ciphertext_length);
+		seed = chunk_alloc(seed_length);
+		drbg->generate(drbg, seed.len, seed.ptr);
+		if (!EVP_PKEY_encapsulate_deterministic(ctx, this->ciphertext.ptr,
+																						&this->ciphertext.len,
+																						this->shared_secret.ptr,
+																						&this->shared_secret.len, seed.ptr,
+																						&seed.len))
+		{
+			EVP_PKEY_free(pkey);
+			EVP_PKEY_CTX_free(ctx);
+			chunk_clear(&this->shared_secret);
+			chunk_clear(&this->ciphertext);
+			chunk_clear(&seed);
+			return FALSE;
+		}
+	}
+	else
+	{
+		EVP_PKEY_encapsulate(ctx, NULL, &ciphertext_length, NULL,
+												 &shared_secret_length);
+		this->shared_secret = chunk_alloc(shared_secret_length);
+		this->ciphertext = chunk_alloc(ciphertext_length);
+		if (!EVP_PKEY_encapsulate(ctx, this->ciphertext.ptr, &this->ciphertext.len,
+															this->shared_secret.ptr,
+															&this->shared_secret.len))
+		{
+			EVP_PKEY_free(pkey);
+			EVP_PKEY_CTX_free(ctx);
+			chunk_clear(&this->shared_secret);
+			chunk_clear(&this->ciphertext);
+			return FALSE;
+		}
+	}
+	EVP_PKEY_free(pkey);
+	EVP_PKEY_CTX_free(ctx);
+	chunk_clear(&seed);
+	return TRUE;
+}
+
+METHOD(key_exchange_t, set_public_key, bool,
+			 private_openssl_kem_t *this, chunk_t value)
+{
+
+	if (this->shared_secret.ptr != NULL || this->ciphertext.ptr != NULL)
+	{
+		return FALSE;
+	}
+
+	/* initiator action */
+	if (this->pkey)
+	{
+		return openssl_kem_decapsulate(this, value);
+	}
+
+	/* responder action */
+	return openssl_kem_encapsulate(this, value);
+}
+
+METHOD(key_exchange_t, set_seed, bool,
+			 private_openssl_kem_t *this, chunk_t ignore, drbg_t *seed)
+{
+	drbg_t *drbg = this->thread_drbg->get(this->thread_drbg);
+	if (!seed)
+	{
+		return FALSE;
+	}
+	DESTROY_IF(drbg);
+	this->thread_drbg->set(this->thread_drbg, seed->get_ref(seed));
+	return TRUE;
+}
+
+METHOD(key_exchange_t, destroy, void,
+private_openssl_kem_t *this)
+{
+	drbg_t *drbg = this->thread_drbg->get(this->thread_drbg);
+	EVP_PKEY_free(this->pkey);
+	chunk_clear(&this->shared_secret);
+	chunk_clear(&this->ciphertext);
+	DESTROY_IF(drbg);
+	this->thread_drbg->destroy(this->thread_drbg);
+	free(this);
+}
+
+/*
+ * Described in header.
+ */
+openssl_kem_t *openssl_kem_create(key_exchange_method_t method)
+{
+	private_openssl_kem_t *this;
+	INIT(this,
+		 .public = {
+			 .ke = {
```

Comment:
As commented in the struct definition, `.ke` can be dropped and `.public` would get the methods assigned directly.